### PR TITLE
Improvements to Video Interface related functions and data

### PR DESCRIPTION
**Pull request by Thar0**
Largely concerns the VI functions in libultra and `z_vimode.c`.
Also some improvements to `devmgr.c` since I replaced `OSMgrArgs` structure with the original `OSDevMgr`.

I couldn't think of short enough names for the last two `ViModeEditState`s, suggestions welcome.

### CODE REVIEW

**Review by fig02 — APPROVED** [review]
the actual content of this code is way over my head so I cant weigh in on any names, but it looks fine to me from what I can tell

File: include/z64.h
Comment (engineer124) [inline include/z64.h:1772]:
```suggestion
typedef enum {
    /* 0 */ VI_MODE_EDIT_STATE_INACTIVE,
    /* 1 */ VI_MODE_EDIT_STATE_ACTIVE,
    /* 2 */ VI_MODE_EDIT_STATE_2, // active, more adjustments
    /* 3 */ VI_MODE_EDIT_STATE_3  // active, more adjustments, print comparison with NTSC LAN1 mode
} ViModeEditState;
```
Just a small cleanup suggestion

**Review by EllipticEllipsis — COMMENTED** [review]
Nice, this one's been in the back of my mind for a while so it's nice to finally see what it does.

File: src/code/game.c
Comment (fig02) [reply src/code/game.c:245]: i was able to remove one of the U's from game.c in something else im working on. they should all be double checked

File: src/code/z_vimode.c
Comment (Thar0) [reply src/code/z_vimode.c:211]: Yeah this seems like something of a bug. I'm not entirely sure why the width is adjusted in the way it is, there's no documentation stating the video width is restricted to particular multiples. I can make a comment but I'm unsure if I should mark it with a bug tag?

File: src/code/z_vimode.c
Comment (Thar0) [reply src/code/z_vimode.c:39]: This is mentioned in the N64 programming manual section 30.1.3

File: src/code/game.c
Comment (Thar0) [reply src/code/game.c:245]: I tried to remove them from this function however I was unsuccessful

File: src/code/z_vimode.c
Comment (EllipticEllipsis) [reply src/code/z_vimode.c:211]: I'd say it is a bug, although it's annoying that we can't tell what it's meant to do.

**Review by Dragorn421 — COMMENTED** [review]
Partial review

Are the VI regs defines like `#define HSYNC` original work or where do they come from?

**Comment by Thar0** [comment]
The VI reg defines are from libreultra & ultralib, not sure if they existed anywhere prior to libreultra

File: src/libultra/io/visetspecial.c
Comment (EllipticEllipsis) [inline src/libultra/io/visetspecial.c:13]: You might as well refer to the place in the docs that mentions this here, I think.

**Review by EllipticEllipsis — COMMENTED** [review]
I think you may have missed the variable in idle.c that seems to be the vimode type (`D_80013960`) (also tells us what arg3 of Sched_Init is, maybe?)

(Also, for some reason on N64 that variable is data instead of bss.)

File: src/libultra/io/visetspecial.c
Comment (Thar0) [reply src/libultra/io/visetspecial.c:13]: There is no manual page that discusses this afaik

**Review by fig02 — APPROVED** [review]

**Review by EllipticEllipsis — APPROVED** [review]

File: src/libultra/io/vimodepallan1.c
Comment (Dragorn421) [inline src/libultra/io/vimodepallan1.c:32]: based on usage in `__osViSwapContext` this isn't directly the origin reg value (thankfully), merely an offset (which I have no idea what its purpose is), so it probably shouldn't use `ORIGIN()`

also inspired from `ViMode_Configure`, 640 is 320\*2 (width\*2) ?

File: src/libultra/io/viswapcontext.c
Comment (Dragorn421) [inline src/libultra/io/viswapcontext.c:47]: `ORIGIN(origin)` here or where origin is set above?

File: src/libultra/io/devmgr.c
Comment (Dragorn421) [inline src/libultra/io/devmgr.c:28]: honestly `messageSend` as a boolean is more confusing to me than a 0/1 or is there something I'm missing

File: include/ultra64/rcp.h
Comment (Dragorn421) [inline include/ultra64/rcp.h:56]: n64brew only mentions 0b11 and the code only mentions 0b11 so probably these two can go?

File: include/ultra64/rcp.h
Comment (Dragorn421) [inline include/ultra64/rcp.h:58]: according to current n64brew docs this is "dedither" not "dither" :shrug:

File: include/ultra64/rcp.h
Comment (Dragorn421) [inline include/ultra64/rcp.h:172]: Only thing I found on this:

> rcp.h
> Changed 8-14 for DEVICE_TYPE to reserve. 

(in "Release Notes OS2.0K" `n64man/relnote_k/index.htm`)

do you know more lol

File: src/libultra/io/visetspecial.c
Comment (EllipticEllipsis) [reply src/libultra/io/visetspecial.c:44]: But is that actually an improvement? We know that frequently they set flags one at a time, this just makes it harder to read what's going on.

Comment (Dragorn421) [reply src/libultra/io/visetspecial.c:44]: I'm not sure what you mean
like,
```c
        __osViNext->features &= ~VI_CTRL_ANTIALIAS_MASK;
        __osViNext->features |= VI_CTRL_ANTIALIAS_MODE_0;
```
doesn't match

Comment (EllipticEllipsis) [reply src/libultra/io/visetspecial.c:44]: I obviously don't understand what you're intending with this, then.

Comment (Dragorn421) [reply src/libultra/io/visetspecial.c:44]: Well the goal of this line is to set the "antialias mode" to `VI_CTRL_ANTIALIAS_MODE_0`, so for searchability/shiftability purposes (even though this relies on N64 hardware so it's not like there's a point in practice) it would be better to reference that name

Comment (EllipticEllipsis) [reply src/libultra/io/visetspecial.c:44]: Hmm, seems like this doesn't actually make sense anyway since it's using it like an enum from the looks of it: reading the description, 3 is not actually 1 | 2 even though it looks like it?

Comment (Dragorn421) [reply src/libultra/io/visetspecial.c:44]: Yeah VI_CTRL_ANTIALIAS_MODE_0/1/2/3 is an enum indeed

File: src/libultra/io/vimodepallan1.c
Comment (Thar0) [reply src/libultra/io/vimodepallan1.c:32]: `ORIGIN()` is for `OSViFieldRegs::origin`, not `VI_ORIGIN_REG`

If we wanted we could make `ORIGIN` be a macro in terms of width, bitdepth and target resolution:
`#define ORIGIN(width, fb16Bit, loRes) ((width) * 2 * ((fb16Bit) ? 1 : 2) * ((loRes) ? 1 : 2)))`
If it were just in terms of `width` it would not work for other VI modes that are not included in oot, breaking compatibility. I'm not sure if it's worth the added complexity

File: src/libultra/io/devmgr.c
Comment (Thar0) [reply src/libultra/io/devmgr.c:28]: I can see that it can be a bit confusing since it was written weirdly, (`messageSend != true`) but I don't think that disqualifies it from being a boolean?

File: include/ultra64/rcp.h
Comment (Thar0) [reply include/ultra64/rcp.h:56]: These other modes do exist to some capacity, afaik, so I think it's better to maintain a complete set of defines.
Also notably, the iQue player needs different pixel advance settings or the output is corrupted

File: include/ultra64/rcp.h
Comment (Thar0) [reply include/ultra64/rcp.h:58]: Right, what it means by "dither filter" is a filter that smooths out dithering applied by earlier stages rather than a filter that performs dithering. Unfortunately, that's just how SGI named it.

File: include/ultra64/rcp.h
Comment (Thar0) [reply include/ultra64/rcp.h:172]: The comments are just lifted straight from OS2.0L rcp.h

File: src/libultra/io/devmgr.c
Comment (Dragorn421) [reply src/libultra/io/devmgr.c:28]: well usually we still boolify `== true`-using stuff when it's otherwise mostly used as a bool, here the only use is `== true`, but whatever

My comment is more that, what does "messageSend == true" mean? the message will be / has been sent ? idk it just looks like a "exit loop" flag to me and for some reason using true/false makes it harder to read

but if it's just me whatever it's not like this is important

File: include/ultra64/rcp.h
Comment (Dragorn421) [reply include/ultra64/rcp.h:56]: why only the two low bits then?

File: src/libultra/io/vimodepallan1.c
Comment (Dragorn421) [reply src/libultra/io/vimodepallan1.c:32]: It would be good to comment on the macros what their purpose is then (if not use prefixes), since "origin" is also a vi reg and the macros look just like the vi reg macros family WIDTH, HSYNC etc

File: include/ultra64/rcp.h
Comment (Dragorn421) [reply include/ultra64/rcp.h:58]: I forgot/didn't notice some of these are original names okay

File: include/ultra64/rcp.h
Comment (Dragorn421) [reply include/ultra64/rcp.h:172]: so rcp.htm in the manual is out-of-date and I guess others too
I'll flee those in the future

File: include/ultra64/rcp.h
Comment (Thar0) [reply include/ultra64/rcp.h]: Because I copied them from ultralib and never noticed it wasn't complete 😎 
I've changed it to one macro with an argument to generate all values.

File: src/libultra/io/visetspecial.c
Comment (Thar0) [reply src/libultra/io/visetspecial.c:44]: I'm reluctant to do this since it's very possible this won't match other libultra versions. (I'm also concerned about matching ultralib, particularly debug builds)

Comment (Dragorn421) [reply src/libultra/io/visetspecial.c:44]: Ok
`// VI_CTRL_ANTIALIAS_MODE_0` ? (don't have to)

File: src/code/z_vimode.c
Comment (Dragorn421) [inline src/code/z_vimode.c:58]: "N=(normal interlacing)" contradicts `loResDeinterlaced = loRes && modeN` below?

**Review by Dragorn421 — APPROVED** [review]
Good to merge!

File: src/code/z_vimode.c
Comment (AngheloAlf) [inline src/code/z_vimode.c:102]: Any reason to not merge this into an `else if` like it was?

Comment (Thar0) [reply src/code/z_vimode.c:102]: The conditions being checked are of a different character (full mode vs only AA on/off), I think it's a bit clearer to see what's going on this way

**Pull request merged**
